### PR TITLE
CODE: Use custom LLVM build, improve clang-format configuration

### PR DESCRIPTION
**Pull request by gleon99**
## What
Use a custom build of `clang-format` when formatting code in the pipeline, which fixes pointer alignment and supports alignment of consecutive declarations even if there is a comment between them.

## Why ?
Improve the automatic code formatting done in the CI, by making it consistent and closer to our coding style, so that making it a mandatory step would make sense.

### CODE REVIEW

File: buildlib/fedora.Dockerfile
Comment (yosefe) [inline buildlib/fedora.Dockerfile:29]: i don't see the clang-format changes in this llvm fork

File: buildlib/fedora.Dockerfile
Comment (yosefe) [inline buildlib/fedora.Dockerfile:32]: DCMAKE_C_COMPILER=gcc probably not needed in docker

File: buildlib/fedora.Dockerfile
Comment (yosefe) [inline buildlib/fedora.Dockerfile:27]: why needed? isn't it created by make install?

File: buildlib/fedora.Dockerfile
Comment (gleon99) [reply buildlib/fedora.Dockerfile:29]: The changes are there, in the master branch:
https://github.com/gleon99/llvm-project/commit/e893bbd9efe3eb29c97548a9e268a1eb625c2bb3

File: .clang-format
Comment (yosefe) [inline .clang-format:62]: pls add UCS_INIT_ONCE

File: buildlib/az-helpers.sh
Comment (yosefe) [inline buildlib/az-helpers.sh:62]: can you pls push bad C file to this PR as temp example of how "SucceededWithIssues" looks?

**Comment by gleon99** [comment]
bot:pipe:retest

File: buildlib/fedora.Dockerfile
Comment (shamisp) [inline buildlib/fedora.Dockerfile:28]: Please keep the llvm-project.git under openucx project

Comment (shamisp) [reply buildlib/fedora.Dockerfile:28]: Other option is directly git clone it from llvm tree (specific version that you patches) and apply our patch from contrtrib or build lib.

**Review by shamisp — CHANGES_REQUESTED** [review]

File: buildlib/fedora.Dockerfile
Comment (gleon99) [reply buildlib/fedora.Dockerfile:28]: @yosefe, that's the exact opposite of what you told me to do (fork to my account and commit there).

Comment (yosefe) [reply buildlib/fedora.Dockerfile:28]: @shamisp I think forking it in OpenUCX org is a better option than keeping patches inside UCX

**Review by yosefe — COMMENTED** [review]
can we also disable SortIncludes for now?
till we figure out a good method for IncludeCategories

File: buildlib/fedora.Dockerfile
Comment (shamisp) [reply buildlib/fedora.Dockerfile:28]: Is there any chance to upstream your changes to LLVM ? This is will help in a long term.

Comment (shamisp) [reply buildlib/fedora.Dockerfile:28]: @gleon99 please work with @yosefe create work under openucx organization. thanks !

**Comment by brminich** [comment]
issue is #6011

bot:pipe:retest

**Review by yosefe — COMMENTED** [review]
can you pls add bad C file for example for format fail?

**Comment by gleon99** [comment]
@yosefe , added format error in `bitmap.h` (assuming we're reverting it before merge).

**Comment by yosefe** [comment]
@gleon99 looks like some weird error + job does not show "success with issues"
```
2020-12-16T08:37:08.9026249Z [command]/usr/bin/bash --noprofile --norc /__w/_temp/49792ab4-09cd-43a4-b293-b11ed3cec313.sh
2020-12-16T08:37:08.9081875Z Checking code format on diff remotes/origin/master..2d3be2ec4d604d84a88d65e17760cb1348ae48d0
2020-12-16T08:37:08.9888945Z Configuration file(s) do(es) not support C++: /__w/1/s/.clang-format
2020-12-16T08:37:08.9897580Z error: `clang-format -lines=193:193 -assume-filename=src/ucs/datastruct/bitmap.h` failed
2020-12-16T08:37:08.9997926Z Generated patch file:
2020-12-16T08:37:09.0042350Z error: unrecognized input
2020-12-16T08:37:09.0627586Z ##[section]Finishing: Bash
```

**Comment by gleon99** [comment]
@yosefe , hopefully fixed. I think `master` is currently broken (the `.clang-format` file). A `---` separator has to be put between language sections.

**Comment by yosefe** [comment]
@gleon99 need to push merge commit with master and then push this fix.. or squash and rebase everything

**Comment by gleon99** [comment]
Rebased over master & squashed everything. Don't merge yet, the incorrectly formatted file is still there.

**Comment by yosefe** [comment]
few things are still off.
maybe the patch for clang is mis-applied?

1.
```
--- a/test/apps/test_cuda_hook.c
+++ b/test/apps/test_cuda_hook.c
@@ -15,7 +15,7 @@
 
 static void event_cb(ucm_event_type_t event_type, ucm_event_t *event, void *arg)
 {
-    int *count_p = arg;
+    int        *count_p = arg;
     const char *title;
 
     if (event_type == UCM_EVENT_MEM_TYPE_ALLOC) {
```

2.
```
@@ -37,11 +37,11 @@ int main(int argc, char **argv)
 {
     static const ucm_event_type_t memtype_events = UCM_EVENT_MEM_TYPE_ALLOC |
                                                    UCM_EVENT_MEM_TYPE_FREE;
-    static const int num_expected_events         = 2;
-    ucs_status_t status;
-    ucp_context_h context;
-    ucp_params_t params;
-    int num_events;
+    static const int num_expected_events = 2;
+    ucs_status_t     status;
+    ucp_context_h    context;
+    ucp_params_t     params;
+    int              num_events;
```

3. SortIncludes is true for Cpp (should be false)

**Comment by gleon99** [comment]
@yosefe ,
(1) - That's the correct and expected behavior when aligning declarations.

(2) - The patch doesn't deal with that case. The thing there is the
```
static const ucm_event_type_t memtype_events = UCM_EVENT_MEM_TYPE_ALLOC |
                                                    UCM_EVENT_MEM_TYPE_FREE;
```
line which breaks the assignment alignment. If it was a single line i.e
```
static const ucm_event_type_t memtype_events = UCM_EVENT_MEM_TYPE_ALLOC;
```

instead, everything would work fine (checked). Do we want to fix that in the patch?

(3) - Fixed.

(4) - That's legitimate behaviour as well (`PointerAlignment: Right`), unless we decide to fix it too.

(5) - Fixed.

**Comment by yosefe** [comment]
@gleon99 ok, we can improve it more later on. pls sqaush and remove the badly formatted file and then we can merge it

**Comment by gleon99** [comment]
@yosefe , removed the misformatted file and squashed.

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
bot:pipe:retest

**Pull request merged**
